# IndiaKart E- Commerce Business Analytics Project 
## Phase 1 - Data Cleaning & Quality Check

**Analyst:** Kartik Jaswal 
### Objective
The objective of this phase is to load, inspect, clean, and validate the IndiaKart datasets before performing business analysis.

The data quality assessment will include:
- Dataset structure and data types
- Missing value analysis
- Duplicate detection
- Date and numeric field validation
- Outlier detection
- Referential integrity checks
- Business-specific data quality checks

## Step 1: Import Libraries and Load Datasets

In this step, the required Python libraries are imported and all eight IndiaKart datasets are loaded into pandas DataFrames for further inspection and analysis.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

print("Libraries imported successfully.")

Libraries imported successfully.



### Extracting the Raw Dataset

The IndiaKart dataset was provided as a compressed ZIP file. The archive was extracted into a dedicated project folder before loading the individual CSV files into Python.

In [3]:
import zipfile
import os

zip_path = r"C:\Users\karti\Desktop\IndiaEKART.zip"
extract_path = r"C:\Users\karti\Desktop\IndiaEKART"

os.makedirs(extract_path, exist_ok=True)

with zipfile.ZipFile(zip_path, "r") as zip_ref:
    zip_ref.extractall(extract_path)

print("ZIP file extracted successfully.")

ZIP file extracted successfully.


In [4]:
for root, dirs, files in os.walk(extract_path):
    for file in files:
        print(os.path.join(root, file))

C:\Users\karti\Desktop\IndiaEKART\customers DA.csv
C:\Users\karti\Desktop\IndiaEKART\data_dictionary __ fin.md
C:\Users\karti\Desktop\IndiaEKART\IndiaKart_Project_PRD__.docx
C:\Users\karti\Desktop\IndiaEKART\inventory DA.csv
C:\Users\karti\Desktop\IndiaEKART\orders DA.csv
C:\Users\karti\Desktop\IndiaEKART\order_items DA.csv
C:\Users\karti\Desktop\IndiaEKART\payments DA.csv
C:\Users\karti\Desktop\IndiaEKART\products DA.csv
C:\Users\karti\Desktop\IndiaEKART\returns DA.csv
C:\Users\karti\Desktop\IndiaEKART\schema_and_sql final internmo.sql
C:\Users\karti\Desktop\IndiaEKART\suppliers DA.csv


## Step 2: Load All IndiaKart Datasets

All eight CSV files were loaded into separate pandas DataFrames. Keeping the datasets separate at this stage allows each table to be independently profiled and validated before relationships and joins are performed.

In [5]:
data_path = r"C:\Users\karti\Desktop\IndiaEKART"

customers = pd.read_csv(os.path.join(data_path, "customers DA.csv"))
orders = pd.read_csv(os.path.join(data_path, "orders DA.csv"))
order_items = pd.read_csv(os.path.join(data_path, "order_items DA.csv"))
products = pd.read_csv(os.path.join(data_path, "products DA.csv"))
payments = pd.read_csv(os.path.join(data_path, "payments DA.csv"))
returns = pd.read_csv(os.path.join(data_path, "returns DA.csv"))
inventory = pd.read_csv(os.path.join(data_path, "inventory DA.csv"))
suppliers = pd.read_csv(os.path.join(data_path, "suppliers DA.csv"))

print("All 8 datasets loaded successfully.")

All 8 datasets loaded successfully.


In [6]:
datasets = {
    "Customers": customers,
    "Orders": orders,
    "Order Items": order_items,
    "Products": products,
    "Payments": payments,
    "Returns": returns,
    "Inventory": inventory,
    "Suppliers": suppliers
}

for name, df in datasets.items():
    print(f"{name}: {df.shape[0]:,} rows × {df.shape[1]} columns")

Customers: 10,000 rows × 17 columns
Orders: 50,000 rows × 19 columns
Order Items: 100,000 rows × 11 columns
Products: 1,000 rows × 17 columns
Payments: 50,000 rows × 13 columns
Returns: 10,000 rows × 10 columns
Inventory: 1,000 rows × 11 columns
Suppliers: 200 rows × 14 columns


In [7]:
total_records = sum(len(df) for df in datasets.values())

print(f"Total records across all datasets: {total_records:,}")

Total records across all datasets: 222,200


## Step 4: Dataset Structure and Data Type Inspection

Each dataset was inspected to understand its column structure, data types, and sample values. This helps identify columns that require type conversion, cleaning, or further validation before analysis.

In [8]:
for name, df in datasets.items():
    print("=" * 80)
    print(f"{name.upper()}")
    print("=" * 80)
    
    print("\nColumns:")
    print(df.columns.tolist())
    
    print("\nData Types:")
    print(df.dtypes)
    
    print("\nSample Records:")
    display(df.head(3))

CUSTOMERS

Columns:
['customer_id', 'first_name', 'last_name', 'email', 'phone', 'city', 'state', 'pincode', 'gender', 'age', 'segment', 'registration_date', 'last_login_date', 'is_verified', 'is_active', 'total_orders', 'total_spent']

Data Types:
customer_id           object
first_name            object
last_name             object
email                 object
phone                  int64
city                  object
state                 object
pincode                int64
gender                object
age                    int64
segment               object
registration_date     object
last_login_date       object
is_verified            int64
is_active              int64
total_orders           int64
total_spent          float64
dtype: object

Sample Records:


,customer_id,first_name,last_name,email,phone,city,state,pincode,gender,age,segment,registration_date,last_login_date,is_verified,is_active,total_orders,total_spent
0,CUST00001,Kunal,Gupta,kunal.gupta223@gmail.com,9117912889,Guwahati,Assam,314040,Male,54,New,30-07-2024,10-04-2025,0,1,7,1028341.51
1,CUST00002,Aryan,Gupta,aryan.gupta336@yahoo.in,9271550550,Navi Mumbai,Maharashtra,799535,Female,26,Budget,05-07-2023,03-05-2024,1,1,1,20337.39
2,CUST00003,Akash,Chatterjee,akash.chatterjee584@hotmail.com,9559676548,Surat,Gujarat,738196,Female,24,Regular,14-05-2022,31-10-2023,1,1,7,265688.26


ORDERS

Columns:
['order_id', 'customer_id', 'order_date', 'order_time', 'status', 'city', 'state', 'pincode', 'total_amount', 'gst_amount', 'shipping_charge', 'discount_amount', 'final_amount', 'payment_method', 'shipping_partner', 'tracking_id', 'delivered_date', 'is_cod', 'channel']

Data Types:
order_id             object
customer_id          object
order_date           object
order_time           object
status               object
city                 object
state                object
pincode               int64
total_amount        float64
gst_amount          float64
shipping_charge       int64
discount_amount     float64
final_amount        float64
payment_method       object
shipping_partner     object
tracking_id          object
delivered_date       object
is_cod                int64
channel              object
dtype: object

Sample Records:


,order_id,customer_id,order_date,order_time,status,city,state,pincode,total_amount,gst_amount,shipping_charge,discount_amount,final_amount,payment_method,shipping_partner,tracking_id,delivered_date,is_cod,channel
0,ORD000001,CUST02946,22-05-2025,17:10:00,Processing,Pune,Maharashtra,300862,100984.19,16364.64,0,20486.45,100984.19,Cash on Delivery,BlueDart,TRK354572763,NaN,1,App
1,ORD000002,CUST04232,16-10-2023,05:32:00,Delivered,Ludhiana,Punjab,387852,254478.01,41599.80,0,19338.79,254478.01,UPI,DTDC,TRK272838570,19-10-2023,0,Website
2,ORD000003,CUST08078,08-10-2023,12:32:00,Cancelled,Meerut,Uttar Pradesh,294014,32557.06,3562.92,0,696.86,32557.06,Debit Card,Amazon Logistics,TRK338505401,NaN,0,App


ORDER ITEMS

Columns:
['item_id', 'order_id', 'product_id', 'product_name', 'category', 'quantity', 'unit_price', 'gst_rate', 'gst_amount', 'discount_amount', 'total_price']

Data Types:
item_id             object
order_id            object
product_id          object
product_name        object
category            object
quantity             int64
unit_price           int64
gst_rate             int64
gst_amount         float64
discount_amount    float64
total_price        float64
dtype: object

Sample Records:


,item_id,order_id,product_id,product_name,category,quantity,unit_price,gst_rate,gst_amount,discount_amount,total_price
0,ITM0000001,ORD000001,PRD0522,HCL Cricket Bat Ultra,Sports & Fitness,1,40989,12,4918.68,7861.45,38046.23
1,ITM0000002,ORD000001,PRD0094,Berger Router Pro,Electronics,1,62532,18,11255.76,12373.11,61414.65
2,ITM0000003,ORD000001,PRD0613,Voltas Educational Toy Standard,Toys & Baby,1,1585,12,190.20,251.89,1523.31


PRODUCTS

Columns:
['product_id', 'product_name', 'category', 'subcategory', 'brand', 'sku', 'mrp', 'selling_price', 'cost_price', 'gst_rate', 'hsn_code', 'weight_grams', 'supplier_id', 'rating', 'review_count', 'is_active', 'launch_date']

Data Types:
product_id        object
product_name      object
category          object
subcategory       object
brand             object
sku               object
mrp                int64
selling_price      int64
cost_price         int64
gst_rate           int64
hsn_code           int64
weight_grams       int64
supplier_id       object
rating           float64
review_count       int64
is_active          int64
launch_date       object
dtype: object

Sample Records:


,product_id,product_name,category,subcategory,brand,sku,mrp,selling_price,cost_price,gst_rate,hsn_code,weight_grams,supplier_id,rating,review_count,is_active,launch_date
0,PRD0001,Usha Smartphone Elite,Electronics,Smartphone,Usha,SKU-ELE-00001,92028,78958,55738,18,1144,250,SUP0198,4.3,3655,1,29-01-2025
1,PRD0002,Gizmore Smartphone Premium,Electronics,Smartphone,Gizmore,SKU-ELE-00002,91470,65917,48052,18,2857,2000,SUP0140,3.6,2321,1,17-05-2024
2,PRD0003,Fire-Boltt Smartphone Max,Electronics,Smartphone,Fire-Boltt,SKU-ELE-00003,148029,118800,68913,18,2349,250,SUP0113,3.5,4412,1,28-12-2024


PAYMENTS

Columns:
['payment_id', 'order_id', 'customer_id', 'payment_date', 'payment_time', 'payment_method', 'amount', 'status', 'transaction_id', 'bank_name', 'gateway', 'refund_amount', 'refund_date']

Data Types:
payment_id         object
order_id           object
customer_id        object
payment_date       object
payment_time       object
payment_method     object
amount            float64
status             object
transaction_id     object
bank_name          object
gateway            object
refund_amount       int64
refund_date       float64
dtype: object

Sample Records:


,payment_id,order_id,customer_id,payment_date,payment_time,payment_method,amount,status,transaction_id,bank_name,gateway,refund_amount,refund_date
0,PAY000001,ORD000001,CUST02946,22-05-2025,17:10:00,Cash on Delivery,100984.19,Success,TXN5110698474,SBI,Google Pay,0,NaN
1,PAY000002,ORD000002,CUST04232,16-10-2023,05:32:00,UPI,254478.01,Success,TXN4067460377,IndusInd,Google Pay,0,NaN
2,PAY000003,ORD000003,CUST08078,08-10-2023,12:32:00,Debit Card,32557.06,Success,TXN7056710694,Axis Bank,PhonePe,0,NaN


RETURNS

Columns:
['return_id', 'order_id', 'customer_id', 'return_date', 'reason', 'return_amount', 'refund_status', 'refund_date', 'return_condition', 'remarks']

Data Types:
return_id            object
order_id             object
customer_id          object
return_date          object
reason               object
return_amount       float64
refund_status        object
refund_date          object
return_condition     object
remarks              object
dtype: object

Sample Records:


,return_id,order_id,customer_id,return_date,reason,return_amount,refund_status,refund_date,return_condition,remarks
0,RET00001,ORD000015,CUST05470,23-12-2024,Size Issue,24167.33,Pending,30-12-2024,Damaged,Customer initiated return within 7 day window
1,RET00002,ORD000017,CUST07581,01-11-2023,Defective Product,11698.27,Refunded,06-11-2023,Damaged,Customer initiated return within 7 day window
2,RET00003,ORD000038,CUST05691,01-05-2025,Wrong Product Delivered,11147.74,Refunded,06-05-2025,Damaged,Customer initiated return within 7 day window


INVENTORY

Columns:
['inventory_id', 'product_id', 'warehouse_location', 'quantity_available', 'quantity_reserved', 'reorder_level', 'reorder_quantity', 'last_restocked_date', 'unit_cost', 'total_inventory_value', 'status']

Data Types:
inventory_id             object
product_id               object
warehouse_location       object
quantity_available        int64
quantity_reserved         int64
reorder_level             int64
reorder_quantity          int64
last_restocked_date      object
unit_cost                 int64
total_inventory_value     int64
status                   object
dtype: object

Sample Records:


,inventory_id,product_id,warehouse_location,quantity_available,quantity_reserved,reorder_level,reorder_quantity,last_restocked_date,unit_cost,total_inventory_value,status
0,INV0001,PRD0001,Hyderabad-WH4,13,4,49,245,11-09-2023,55738,724594,Low Stock
1,INV0002,PRD0002,Kolkata-WH6,368,14,50,250,10-03-2025,48052,17683136,In Stock
2,INV0003,PRD0003,Delhi-WH2,196,19,44,220,22-11-2024,68913,13506948,In Stock


SUPPLIERS

Columns:
['supplier_id', 'supplier_name', 'contact_person', 'email', 'phone', 'city', 'state', 'pincode', 'category', 'gstin', 'payment_terms_days', 'rating', 'created_date', 'is_active']

Data Types:
supplier_id            object
supplier_name          object
contact_person         object
email                  object
phone                   int64
city                   object
state                  object
pincode                 int64
category               object
gstin                  object
payment_terms_days      int64
rating                float64
created_date           object
is_active               int64
dtype: object

Sample Records:


,supplier_id,supplier_name,contact_person,email,phone,city,state,pincode,category,gstin,payment_terms_days,rating,created_date,is_active
0,SUP0001,Iyer Elec Pvt Ltd,Harish Iyer,contact@iyerele.com,9362950628,Delhi,Delhi,334053,Electronics,14TRXCK1488C4Z9,15,4.1,22-05-2023,1
1,SUP0002,Sharma Offi Pvt Ltd,Mohit Sharma,contact@sharmaoff.com,9914763202,Ahmedabad,Gujarat,267414,Office Supplies,32KHFTC7224C6Z5,15,4.5,25-03-2023,1
2,SUP0003,Choudhury Beau Pvt Ltd,Nandini Choudhury,contact@choudhurybea.com,9764130526,Pune,Maharashtra,479201,Beauty & Health,28EBRUZ4814W7Z5,60,4.3,27-12-2022,1


In [9]:
print("ORDERS")
print("=" * 80)

print("\nColumns:")
print(orders.columns.tolist())

print("\nData Types:")
print(orders.dtypes)

print("\nSample Records:")
display(orders.head(5))

ORDERS

Columns:
['order_id', 'customer_id', 'order_date', 'order_time', 'status', 'city', 'state', 'pincode', 'total_amount', 'gst_amount', 'shipping_charge', 'discount_amount', 'final_amount', 'payment_method', 'shipping_partner', 'tracking_id', 'delivered_date', 'is_cod', 'channel']

Data Types:
order_id             object
customer_id          object
order_date           object
order_time           object
status               object
city                 object
state                object
pincode               int64
total_amount        float64
gst_amount          float64
shipping_charge       int64
discount_amount     float64
final_amount        float64
payment_method       object
shipping_partner     object
tracking_id          object
delivered_date       object
is_cod                int64
channel              object
dtype: object

Sample Records:


,order_id,customer_id,order_date,order_time,status,city,state,pincode,total_amount,gst_amount,shipping_charge,discount_amount,final_amount,payment_method,shipping_partner,tracking_id,delivered_date,is_cod,channel
0,ORD000001,CUST02946,22-05-2025,17:10:00,Processing,Pune,Maharashtra,300862,100984.19,16364.64,0,20486.45,100984.19,Cash on Delivery,BlueDart,TRK354572763,NaN,1,App
1,ORD000002,CUST04232,16-10-2023,05:32:00,Delivered,Ludhiana,Punjab,387852,254478.01,41599.80,0,19338.79,254478.01,UPI,DTDC,TRK272838570,19-10-2023,0,Website
2,ORD000003,CUST08078,08-10-2023,12:32:00,Cancelled,Meerut,Uttar Pradesh,294014,32557.06,3562.92,0,696.86,32557.06,Debit Card,Amazon Logistics,TRK338505401,NaN,0,App
3,ORD000004,CUST09612,02-10-2023,08:05:00,Delivered,Nashik,Maharashtra,626047,24836.94,3092.88,0,4029.94,24836.94,UPI,DTDC,TRK289792017,09-10-2023,0,Mobile Web
4,ORD000005,CUST03065,15-06-2024,20:19:00,Delivered,Moradabad,Uttar Pradesh,409210,156689.29,25092.36,0,7805.07,156689.29,UPI,Ecom Express,TRK978240041,22-06-2024,0,App


## Step 5: Date Type Conversion — Orders

The `order_date` and `delivered_date` columns were initially stored as text (object) rather than datetime values.

Both columns were converted to pandas datetime format using the DD-MM-YYYY structure. Missing delivery dates were retained because orders that are cancelled, processing, or not yet delivered may legitimately have no delivery date.

In [10]:
orders['order_date'] = pd.to_datetime(
    orders['order_date'],
    format='%d-%m-%Y',
    errors='coerce'
)

orders['delivered_date'] = pd.to_datetime(
    orders['delivered_date'],
    format='%d-%m-%Y',
    errors='coerce'
)

print(orders[['order_date', 'delivered_date']].dtypes)

order_date        datetime64[ns]
delivered_date    datetime64[ns]
dtype: object


In [11]:
print("Missing/Invalid order dates:")
print(orders['order_date'].isna().sum())

print("\nMissing delivery dates:")
print(orders['delivered_date'].isna().sum())

Missing/Invalid order dates:
0

Missing delivery dates:
17501


In [12]:
print("Order Date Range:")
print("Earliest:", orders['order_date'].min())
print("Latest:", orders['order_date'].max())

print("\nDelivered Date Range:")
print("Earliest:", orders['delivered_date'].min())
print("Latest:", orders['delivered_date'].max())

Order Date Range:
Earliest: 2023-06-24 00:00:00
Latest: 2025-06-23 00:00:00

Delivered Date Range:
Earliest: 2023-06-26 00:00:00
Latest: 2025-06-30 00:00:00


## Step 6: Delivery Date Validation

Delivery dates were validated against order dates to identify logically inconsistent records.

A valid delivered order should have a delivery date on or after its order date. Missing delivery dates were also analysed by order status to distinguish genuine data quality issues from valid business cases.

In [13]:
invalid_delivery_dates = orders[
    orders['delivered_date'].notna() &
    (orders['delivered_date'] < orders['order_date'])
]

print("Orders where delivered_date is before order_date:")
print(len(invalid_delivery_dates))

Orders where delivered_date is before order_date:
0


In [14]:
delivered_missing_date = orders[
    (orders['status'] == 'Delivered') &
    (orders['delivered_date'].isna())
]

print(
    "Delivered orders with missing delivered_date:",
    len(delivered_missing_date))

Delivered orders with missing delivered_date: 0


In [15]:
missing_delivery_by_status = (
    orders[orders['delivered_date'].isna()]
    ['status']
    .value_counts()
)

print("Missing delivery dates by order status:")
print(missing_delivery_by_status)

Missing delivery dates by order status:
status
Cancelled     5996
Shipped       5071
Returned      3933
Processing    2501
Name: count, dtype: int64


### Observation

- No records were found where `delivered_date` occurred before `order_date`.
- All orders with a `Delivered` status have a valid delivery date.
- Missing delivery dates occur only among non-delivered order statuses and were therefore retained rather than imputed.
- Returned orders with missing delivery dates will be reviewed further during cross-table validation with the returns dataset.

## Step 7: Duplicate Validation — Orders

Duplicate records and duplicate order IDs were checked to ensure that each order is uniquely represented in the dataset.

In [16]:
duplicate_rows = orders.duplicated().sum()

print("Completely duplicate rows:", duplicate_rows)

Completely duplicate rows: 0


In [17]:
duplicate_order_ids = orders['order_id'].duplicated().sum()

print("Duplicate order IDs:", duplicate_order_ids)

Duplicate order IDs: 0


In [18]:
print("Total rows:", len(orders))
print("Unique order IDs:", orders['order_id'].nunique())

Total rows: 50000
Unique order IDs: 50000


In [19]:
high_value_orders = orders[
    orders['final_amount'] > 500000
]

print("Orders above ₹5,00,000:", len(high_value_orders))

display(
    high_value_orders[
        [
            'order_id',
            'customer_id',
            'order_date',
            'status',
            'final_amount',
            'payment_method',
            'state'
        ]
    ].head(10)
)

Orders above ₹5,00,000: 312


,order_id,customer_id,order_date,status,final_amount,payment_method,state
33,ORD000034,CUST01867,2024-03-10,Delivered,527178.05,Credit Card,Madhya Pradesh
34,ORD000035,CUST01370,2025-05-26,Cancelled,532061.13,UPI,Maharashtra
306,ORD000307,CUST03874,2024-03-20,Cancelled,550716.17,Cash on Delivery,Chhattisgarh
583,ORD000584,CUST07624,2024-10-30,Delivered,745667.68,Cash on Delivery,Telangana
590,ORD000591,CUST05298,2025-05-22,Cancelled,520293.72,Credit Card,Haryana
763,ORD000764,CUST00862,2025-04-07,Delivered,540156.71,UPI,Punjab
985,ORD000986,CUST06894,2024-11-27,Shipped,512079.51,Wallet,Uttar Pradesh
1092,ORD001093,CUST00492,2025-04-28,Cancelled,680009.23,UPI,Delhi
1114,ORD001115,CUST03544,2025-01-04,Delivered,646223.50,Debit Card,Maharashtra
1178,ORD001179,CUST07815,2025-05-15,Processing,513624.71,Credit Card,Uttar Pradesh


### Duplicate and High-Value Order Findings

- No completely duplicated records were found in the orders dataset.
- All 50,000 order IDs are unique, confirming that `order_id` behaves as a valid primary key.
- 312 orders have a final amount greater than ₹5,00,000.
- These high-value orders were flagged for further investigation rather than removed automatically, because a high transaction value does not necessarily indicate incorrect data.

In [20]:
orders['final_amount'].describe()

count    5.000000e+04
mean     6.294771e+04
std      9.789750e+04
min      1.360700e+02
25%      6.893520e+03
50%      2.418338e+04
75%      7.149574e+04
max      1.133382e+06
Name: final_amount, dtype: float64

In [21]:
Q1 = orders['final_amount'].quantile(0.25)
Q3 = orders['final_amount'].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

print(f"Q1: ₹{Q1:,.2f}")
print(f"Q3: ₹{Q3:,.2f}")
print(f"IQR: ₹{IQR:,.2f}")
print(f"Lower Limit: ₹{lower_limit:,.2f}")
print(f"Upper Limit: ₹{upper_limit:,.2f}")

Q1: ₹6,893.52
Q3: ₹71,495.74
IQR: ₹64,602.22
Lower Limit: ₹-90,009.81
Upper Limit: ₹168,399.07


In [22]:
final_amount_outliers = orders[
    (orders['final_amount'] < lower_limit) |
    (orders['final_amount'] > upper_limit)
]

print("Total IQR outliers:", len(final_amount_outliers))

print(
    "Outlier percentage:",
    round(len(final_amount_outliers) / len(orders) * 100, 2),
    "%"
)

Total IQR outliers: 5243
Outlier percentage: 10.49 %


In [23]:
print(f"Minimum Final Amount: ₹{orders['final_amount'].min():,.2f}")
print(f"Maximum Final Amount: ₹{orders['final_amount'].max():,.2f}")
print(f"Mean Final Amount: ₹{orders['final_amount'].mean():,.2f}")
print(f"Median Final Amount: ₹{orders['final_amount'].median():,.2f}")

Minimum Final Amount: ₹136.07
Maximum Final Amount: ₹1,133,382.48
Mean Final Amount: ₹62,947.71
Median Final Amount: ₹24,183.38


### Final Amount Outlier Analysis

The distribution of order values is strongly right-skewed. The mean order value (₹62,947.71) is considerably higher than the median (₹24,183.38), indicating the presence of high-value transactions.

Using the IQR method:
- Q1: ₹6,893.52
- Q3: ₹71,495.74
- IQR: ₹64,602.22
- Upper outlier threshold: ₹168,399.07
- IQR outliers: 5,243 orders (10.49%)
- Orders above the business review threshold of ₹5,00,000: 312
- Maximum observed order value: ₹1,133,382.48

These records were flagged but not removed because high-value transactions can be legitimate in an e-commerce marketplace. They will be retained for analysis unless further validation identifies data-quality problems.

## Step 10: Missing Value Analysis

All eight datasets were checked for missing values. Both the number and percentage of missing records were calculated for each affected column.

Missing values are not automatically treated as errors. Their business meaning is evaluated before deciding whether to retain, remove, or impute them.

In [24]:
for name, df in datasets.items():
    
    missing_count = df.isnull().sum()
    missing_pct = (df.isnull().sum() / len(df) * 100).round(2)

    missing_summary = pd.DataFrame({
        'Missing_Count': missing_count,
        'Missing_Percentage': missing_pct
    })

    missing_summary = missing_summary[
        missing_summary['Missing_Count'] > 0
    ]

    print("\n" + "=" * 70)
    print(f"{name.upper()} — MISSING VALUES")
    print("=" * 70)

    if missing_summary.empty:
        print("No missing values found.")
    else:
        display(missing_summary)


CUSTOMERS — MISSING VALUES
No missing values found.

ORDERS — MISSING VALUES


,Missing_Count,Missing_Percentage
delivered_date,17501,35.0



ORDER ITEMS — MISSING VALUES
No missing values found.

PRODUCTS — MISSING VALUES
No missing values found.

PAYMENTS — MISSING VALUES


,Missing_Count,Missing_Percentage
refund_date,50000,100.0



RETURNS — MISSING VALUES
No missing values found.

INVENTORY — MISSING VALUES
No missing values found.

SUPPLIERS — MISSING VALUES
No missing values found.


### Missing Value Findings

The overall dataset is highly complete. Six of the eight tables contain no missing values.

- `orders.delivered_date` contains 17,501 missing values (35.00%). These were retained because the missing values occur among non-delivered order statuses and therefore can represent valid business conditions.
- `payments.refund_date` is missing for all 50,000 payment records (100%). This requires further investigation before deciding whether the column should be removed.
- No missing-value imputation was performed without first establishing the business meaning of the missing data.

In [25]:
print("PAYMENTS COLUMNS")
print(payments.columns.tolist())

print("\nDATA TYPES")
print(payments.dtypes)

display(payments.head())

PAYMENTS COLUMNS
['payment_id', 'order_id', 'customer_id', 'payment_date', 'payment_time', 'payment_method', 'amount', 'status', 'transaction_id', 'bank_name', 'gateway', 'refund_amount', 'refund_date']

DATA TYPES
payment_id         object
order_id           object
customer_id        object
payment_date       object
payment_time       object
payment_method     object
amount            float64
status             object
transaction_id     object
bank_name          object
gateway            object
refund_amount       int64
refund_date       float64
dtype: object


,payment_id,order_id,customer_id,payment_date,payment_time,payment_method,amount,status,transaction_id,bank_name,gateway,refund_amount,refund_date
0,PAY000001,ORD000001,CUST02946,22-05-2025,17:10:00,Cash on Delivery,100984.19,Success,TXN5110698474,SBI,Google Pay,0,NaN
1,PAY000002,ORD000002,CUST04232,16-10-2023,05:32:00,UPI,254478.01,Success,TXN4067460377,IndusInd,Google Pay,0,NaN
2,PAY000003,ORD000003,CUST08078,08-10-2023,12:32:00,Debit Card,32557.06,Success,TXN7056710694,Axis Bank,PhonePe,0,NaN
3,PAY000004,ORD000004,CUST09612,02-10-2023,08:05:00,UPI,24836.94,Success,TXN7460208257,Axis Bank,PhonePe,0,NaN
4,PAY000005,ORD000005,CUST03065,15-06-2024,20:19:00,UPI,156689.29,Success,TXN5835043807,PNB,Razorpay,0,NaN


In [26]:
for col in payments.columns:
    if payments[col].dtype == 'object':
        print(f"\n{col}:")
        print(payments[col].value_counts(dropna=False).head(10))


payment_id:
payment_id
PAY000001    1
PAY033351    1
PAY033329    1
PAY033330    1
PAY033331    1
PAY033332    1
PAY033333    1
PAY033334    1
PAY033335    1
PAY033336    1
Name: count, dtype: int64

order_id:
order_id
ORD000001    1
ORD033351    1
ORD033329    1
ORD033330    1
ORD033331    1
ORD033332    1
ORD033333    1
ORD033334    1
ORD033335    1
ORD033336    1
Name: count, dtype: int64

customer_id:
customer_id
CUST00552    19
CUST05094    19
CUST03174    18
CUST05554    18
CUST07574    18
CUST09398    18
CUST09613    18
CUST03137    17
CUST07425    17
CUST02685    17
Name: count, dtype: int64

payment_date:
payment_date
29-09-2023    98
03-05-2025    91
12-03-2024    91
25-01-2025    91
21-12-2024    90
18-09-2023    89
19-07-2023    87
11-01-2025    87
16-01-2025    87
03-12-2024    87
Name: count, dtype: int64

payment_time:
payment_time
23:27:00    55
18:20:00    53
14:54:00    53
20:01:00    52
04:13:00    51
01:41:00    51
18:34:00    51
13:28:00    51
10:55:00    51
02:51

In [27]:
print("Total payment records:", len(payments))

if 'payment_status' in payments.columns:
    print("\nPayment Status Distribution:")
    print(payments['payment_status'].value_counts(dropna=False))

if 'refund_amount' in payments.columns:
    print("\nRefund Amount Summary:")
    print(payments['refund_amount'].describe())

Total payment records: 50000

Refund Amount Summary:
count    50000.0
mean         0.0
std          0.0
min          0.0
25%          0.0
50%          0.0
75%          0.0
max          0.0
Name: refund_amount, dtype: float64


### Refund Data Investigation

The `refund_date` column contains 100% missing values. Further investigation showed that `refund_amount` is ₹0 for all 50,000 payment records.

This indicates that no refund transaction has been recorded in the payments table. Therefore, the missing `refund_date` values are logically consistent with the available payment data.

No imputation was performed. The column was retained to preserve the original dataset structure.

In [28]:
print(payments.columns.tolist())

['payment_id', 'order_id', 'customer_id', 'payment_date', 'payment_time', 'payment_method', 'amount', 'status', 'transaction_id', 'bank_name', 'gateway', 'refund_amount', 'refund_date']


In [29]:
print("Payment Status Distribution:")
print(payments['status'].value_counts(dropna=False))

Payment Status Distribution:
status
Success    48252
Failed      1748
Name: count, dtype: int64


In [30]:
failed_payments = (payments['status'] == 'Failed').sum()
total_payments = len(payments)

payment_failure_rate = (
    failed_payments / total_payments
) * 100

print(f"Total Payments: {total_payments:,}")
print(f"Failed Payments: {failed_payments:,}")
print(f"Payment Failure Rate: {payment_failure_rate:.2f}%")

Total Payments: 50,000
Failed Payments: 1,748
Payment Failure Rate: 3.50%


### Payment Status and Failure Rate

The payments dataset contains 50,000 payment transactions:

- Successful payments: 48,252
- Failed payments: 1,748
- Payment Failure Rate: 3.50%
- Management benchmark: < 2%

The observed payment failure rate of 3.50% is above the target benchmark of 2%. This represents a potential revenue leakage risk and should be investigated further by payment method, gateway, and bank to identify where failures are concentrated.

No records were removed because failed payments represent valid business events rather than data-quality errors.

## Step 12: Duplicate Check Across All Datasets

All eight datasets were checked for completely duplicated rows. Duplicate records can inflate order counts, revenue, customer metrics, and other business KPIs if they are not identified before analysis.

In [31]:
duplicate_summary = []

for name, df in datasets.items():
    
    duplicate_count = df.duplicated().sum()
    
    duplicate_summary.append({
        'Dataset': name,
        'Total_Rows': len(df),
        'Duplicate_Rows': duplicate_count,
        'Duplicate_Percentage': round(
            duplicate_count / len(df) * 100, 2
        )
    })

duplicate_summary = pd.DataFrame(duplicate_summary)

display(duplicate_summary)

,Dataset,Total_Rows,Duplicate_Rows,Duplicate_Percentage
0,Customers,10000,0,0.0
1,Orders,50000,0,0.0
2,Order Items,100000,0,0.0
3,Products,1000,0,0.0
4,Payments,50000,0,0.0
5,Returns,10000,0,0.0
6,Inventory,1000,0,0.0
7,Suppliers,200,0,0.0


In [32]:
primary_keys = {
    'Customers': 'customer_id',
    'Orders': 'order_id',
    'Order Items': 'item_id',
    'Products': 'product_id',
    'Payments': 'payment_id',
    'Returns': 'return_id',
    'Suppliers': 'supplier_id'
}

for name, key in primary_keys.items():
    
    df = datasets[name]
    
    print(
        f"{name}: "
        f"{len(df):,} rows | "
        f"{df[key].nunique():,} unique {key} | "
        f"{df[key].duplicated().sum():,} duplicate IDs"
    )

Customers: 10,000 rows | 10,000 unique customer_id | 0 duplicate IDs
Orders: 50,000 rows | 50,000 unique order_id | 0 duplicate IDs
Order Items: 100,000 rows | 100,000 unique item_id | 0 duplicate IDs
Products: 1,000 rows | 1,000 unique product_id | 0 duplicate IDs
Payments: 50,000 rows | 50,000 unique payment_id | 0 duplicate IDs
Returns: 10,000 rows | 10,000 unique return_id | 0 duplicate IDs
Suppliers: 200 rows | 200 unique supplier_id | 0 duplicate IDs


### Duplicate Validation Findings

- No completely duplicated rows were found in any of the eight datasets.
- All tested primary identifiers are unique:
  - 10,000 unique customer IDs
  - 50,000 unique order IDs
  - 100,000 unique item IDs
  - 1,000 unique product IDs
  - 50,000 unique payment IDs
  - 10,000 unique return IDs
  - 200 unique supplier IDs
- No duplicate records were removed.
- Inventory was excluded from the primary-key uniqueness test until its warehouse/product structure is validated.

## Step 13: Referential Integrity Validation

Referential integrity checks were performed to confirm that foreign-key values in transactional tables have matching records in their parent tables.

The following relationships were validated:

- Order Items → Orders
- Payments → Orders
- Returns → Orders
- Orders → Customers
- Order Items → Products

Records without a corresponding parent record would be treated as orphan records and flagged as data-quality issues.

In [33]:
integrity_checks = {
    "Order Items → Orders": (
        order_items['order_id'],
        orders['order_id']
    ),
    
    "Payments → Orders": (
        payments['order_id'],
        orders['order_id']
    ),
    
    "Returns → Orders": (
        returns['order_id'],
        orders['order_id']
    ),
    
    "Orders → Customers": (
        orders['customer_id'],
        customers['customer_id']
    ),
    
    "Order Items → Products": (
        order_items['product_id'],
        products['product_id']
    )
}

integrity_results = []

for relationship, (child, parent) in integrity_checks.items():
    
    orphan_count = (~child.isin(parent)).sum()
    
    integrity_results.append({
        "Relationship": relationship,
        "Records_Checked": len(child),
        "Orphan_Records": orphan_count,
        "Integrity_Status": "PASS" if orphan_count == 0 else "FAIL"
    })

integrity_results = pd.DataFrame(integrity_results)

display(integrity_results)

,Relationship,Records_Checked,Orphan_Records,Integrity_Status
0,Order Items → Orders,100000,0,PASS
1,Payments → Orders,50000,0,PASS
2,Returns → Orders,10000,0,PASS
3,Orders → Customers,50000,0,PASS
4,Order Items → Products,100000,0,PASS


### Referential Integrity Findings

All tested relationships passed the referential integrity checks.

- All 100,000 order-item records have matching orders.
- All 50,000 payment records have matching orders.
- All 10,000 return records have matching orders.
- All 50,000 orders are linked to valid customers.
- All 100,000 order-item records reference valid products.
- No orphan records were identified.

This confirms that the key relationships required for cross-table business analysis are structurally consistent.

In [34]:
print("INVENTORY COLUMNS:")
print(inventory.columns.tolist())

print("\nDATA TYPES:")
print(inventory.dtypes)

display(inventory.head())

INVENTORY COLUMNS:
['inventory_id', 'product_id', 'warehouse_location', 'quantity_available', 'quantity_reserved', 'reorder_level', 'reorder_quantity', 'last_restocked_date', 'unit_cost', 'total_inventory_value', 'status']

DATA TYPES:
inventory_id             object
product_id               object
warehouse_location       object
quantity_available        int64
quantity_reserved         int64
reorder_level             int64
reorder_quantity          int64
last_restocked_date      object
unit_cost                 int64
total_inventory_value     int64
status                   object
dtype: object


,inventory_id,product_id,warehouse_location,quantity_available,quantity_reserved,reorder_level,reorder_quantity,last_restocked_date,unit_cost,total_inventory_value,status
0,INV0001,PRD0001,Hyderabad-WH4,13,4,49,245,11-09-2023,55738,724594,Low Stock
1,INV0002,PRD0002,Kolkata-WH6,368,14,50,250,10-03-2025,48052,17683136,In Stock
2,INV0003,PRD0003,Delhi-WH2,196,19,44,220,22-11-2024,68913,13506948,In Stock
3,INV0004,PRD0004,Delhi-WH2,432,10,19,57,10-11-2024,80055,34583760,In Stock
4,INV0005,PRD0005,Bengaluru-WH3,404,16,44,88,05-08-2024,8398,3392792,In Stock


In [35]:
for col in inventory.columns:
    if inventory[col].dtype == 'object':
        print(f"\n{col}:")
        print(inventory[col].value_counts(dropna=False).head(15))


inventory_id:
inventory_id
INV0001    1
INV0672    1
INV0659    1
INV0660    1
INV0661    1
INV0662    1
INV0663    1
INV0664    1
INV0665    1
INV0666    1
INV0667    1
INV0668    1
INV0669    1
INV0670    1
INV0671    1
Name: count, dtype: int64

product_id:
product_id
PRD0001    1
PRD0672    1
PRD0659    1
PRD0660    1
PRD0661    1
PRD0662    1
PRD0663    1
PRD0664    1
PRD0665    1
PRD0666    1
PRD0667    1
PRD0668    1
PRD0669    1
PRD0670    1
PRD0671    1
Name: count, dtype: int64

warehouse_location:
warehouse_location
Chennai-WH5      190
Bengaluru-WH3    171
Hyderabad-WH4    167
Kolkata-WH6      166
Delhi-WH2        165
Mumbai-WH1       141
Name: count, dtype: int64

last_restocked_date:
last_restocked_date
06-10-2023    6
29-09-2024    5
11-08-2024    5
11-07-2023    5
23-01-2024    5
23-05-2025    5
14-01-2024    5
25-11-2023    5
24-06-2023    5
25-04-2024    5
29-01-2024    5
27-04-2024    5
28-01-2025    5
06-08-2023    4
17-01-2025    4
Name: count, dtype: int64

statu

In [36]:
zero_order_customers = customers[
    customers['total_orders'] == 0
]

print("Total Customers:", f"{len(customers):,}")
print("Customers with 0 orders:", f"{len(zero_order_customers):,}")

zero_order_pct = (
    len(zero_order_customers) / len(customers)
) * 100

print(f"Percentage of customers with 0 orders: {zero_order_pct:.2f}%")

Total Customers: 10,000
Customers with 0 orders: 457
Percentage of customers with 0 orders: 4.57%


### Inventory and Customer Validation Findings

**Inventory**
- In Stock: 922 products
- Low Stock: 76 products
- Out of Stock: 2 products
- Only 2 of the 1,000 inventory records are currently classified as Out of Stock.
- Low-stock products should be monitored because they may create future stock availability risks.

**Customers**
- Total customers: 10,000
- Customers with zero orders: 457
- Percentage with zero orders: 4.57%
- Customers with zero orders were retained because they represent valid registered customers who have not yet completed a purchase.

In [37]:
inventory['last_restocked_date'] = pd.to_datetime(
    inventory['last_restocked_date'],
    format='%d-%m-%Y',
    errors='coerce'
)

print(inventory['last_restocked_date'].dtype)
print("Invalid/Missing Restock Dates:",
      inventory['last_restocked_date'].isna().sum())

print("Earliest Restock Date:",
      inventory['last_restocked_date'].min())

print("Latest Restock Date:",
      inventory['last_restocked_date'].max())

datetime64[ns]
Invalid/Missing Restock Dates: 0
Earliest Restock Date: 2023-06-24 00:00:00
Latest Restock Date: 2025-06-22 00:00:00


## Step 17: Customer Order Count Validation

The `customers` table contains a pre-calculated `total_orders` field. To validate this metric, actual order counts were independently calculated from the orders table and compared with the customer-level values.

This check ensures that customer lifetime statistics are consistent with the underlying transaction data.

In [38]:
actual_order_counts = (
    orders.groupby('customer_id')
    .size()
    .reset_index(name='actual_order_count')
)

customer_validation = customers.merge(
    actual_order_counts,
    on='customer_id',
    how='left'
)

customer_validation['actual_order_count'] = (
    customer_validation['actual_order_count']
    .fillna(0)
    .astype(int)
)

customer_validation['order_count_difference'] = (
    customer_validation['total_orders']
    - customer_validation['actual_order_count']
)

mismatched_customers = customer_validation[
    customer_validation['order_count_difference'] != 0
]

print("Total Customers:", len(customer_validation))
print("Matching Order Counts:",
      (customer_validation['order_count_difference'] == 0).sum())

print("Mismatched Order Counts:",
      len(mismatched_customers))

Total Customers: 10000
Matching Order Counts: 10000
Mismatched Order Counts: 0


In [39]:
display(
    mismatched_customers[
        [
            'customer_id',
            'total_orders',
            'actual_order_count',
            'order_count_difference'
        ]
    ].head(10)
)

,customer_id,total_orders,actual_order_count,order_count_difference


### Customer Order Count Validation Result

All customer-level order counts were successfully validated against the orders transaction table.

- Mismatched customer records: 0
- The pre-calculated `total_orders` field is fully consistent with the underlying order transactions.
- Therefore, the field can be reliably used in subsequent customer analysis.

### Inventory Date and Customer Order Validation Findings

**Inventory Restock Dates**
- `last_restocked_date` was successfully converted to datetime format.
- No invalid or missing restock dates were identified.
- Restock dates range from 24 June 2023 to 22 June 2025.

**Customer Order Count Validation**
- All 10,000 customer records were compared with actual transactions in the orders table.
- 10,000 customers had matching order counts.
- No mismatches were identified.
- The pre-calculated `total_orders` field is therefore consistent with the underlying order transaction data.

### Inventory Date and Customer Order Validation Findings

**Inventory Restock Dates**
- `last_restocked_date` was successfully converted to datetime format.
- No invalid or missing restock dates were identified.
- Restock dates range from 24 June 2023 to 22 June 2025.

**Customer Order Count Validation**
- All 10,000 customer records were compared with actual transactions in the orders table.
- 10,000 customers had matching order counts.
- No mismatches were identified.
- The pre-calculated `total_orders` field is therefore consistent with the underlying order transaction data.

In [40]:
for name, df in datasets.items():
    
    date_columns = [
        col for col in df.columns
        if 'date' in col.lower()
    ]
    
    print(f"{name}: {date_columns}")

Customers: ['registration_date', 'last_login_date']
Orders: ['order_date', 'delivered_date']
Order Items: []
Products: ['launch_date']
Payments: ['payment_date', 'refund_date']
Returns: ['return_date', 'refund_date']
Inventory: ['last_restocked_date']
Suppliers: ['created_date']


In [41]:
for name, df in datasets.items():
    
    date_columns = [
        col for col in df.columns
        if 'date' in col.lower()
    ]
    
    if date_columns:
        print(f"\n{name}")
        print(df[date_columns].dtypes)


Customers
registration_date    object
last_login_date      object
dtype: object

Orders
order_date        datetime64[ns]
delivered_date    datetime64[ns]
dtype: object

Products
launch_date    object
dtype: object

Payments
payment_date     object
refund_date     float64
dtype: object

Returns
return_date    object
refund_date    object
dtype: object

Inventory
last_restocked_date    datetime64[ns]
dtype: object

Suppliers
created_date    object
dtype: object


## Step 19: Standardising Date Columns

All remaining date fields stored as text were converted to pandas datetime format using the DD-MM-YYYY structure.

The `payments.refund_date` field was excluded from conversion because all values are missing and no refund transaction is recorded in the payments table.

In [42]:
date_conversion_map = {
    'Customers': {
        'df': customers,
        'columns': ['registration_date', 'last_login_date']
    },
    'Products': {
        'df': products,
        'columns': ['launch_date']
    },
    'Payments': {
        'df': payments,
        'columns': ['payment_date']
    },
    'Returns': {
        'df': returns,
        'columns': ['return_date', 'refund_date']
    },
    'Suppliers': {
        'df': suppliers,
        'columns': ['created_date']
    }
}

for name, info in date_conversion_map.items():
    
    df = info['df']
    
    for col in info['columns']:
        df[col] = pd.to_datetime(
            df[col],
            format='%d-%m-%Y',
            errors='coerce'
        )

print("Date conversion completed.")

Date conversion completed.


In [43]:
for name, info in date_conversion_map.items():
    
    df = info['df']
    
    print("\n" + "=" * 60)
    print(name.upper())
    print("=" * 60)
    
    for col in info['columns']:
        print(
            f"{col}: "
            f"dtype={df[col].dtype}, "
            f"missing/invalid={df[col].isna().sum()}"
        )


CUSTOMERS
registration_date: dtype=datetime64[ns], missing/invalid=0
last_login_date: dtype=datetime64[ns], missing/invalid=0

PRODUCTS
launch_date: dtype=datetime64[ns], missing/invalid=0

PAYMENTS
payment_date: dtype=datetime64[ns], missing/invalid=0

RETURNS
return_date: dtype=datetime64[ns], missing/invalid=0
refund_date: dtype=datetime64[ns], missing/invalid=0

SUPPLIERS
created_date: dtype=datetime64[ns], missing/invalid=0


In [44]:
for name, info in date_conversion_map.items():
    
    df = info['df']
    
    print(f"\n{name}")
    
    for col in info['columns']:
        print(
            f"{col}: "
            f"{df[col].min()} → {df[col].max()}"
        )


Customers
registration_date: 2020-06-24 00:00:00 → 2025-06-23 00:00:00
last_login_date: 2020-07-16 00:00:00 → 2025-06-24 00:00:00

Products
launch_date: 2021-06-24 00:00:00 → 2025-06-23 00:00:00

Payments
payment_date: 2023-06-24 00:00:00 → 2025-06-23 00:00:00

Returns
return_date: 2023-06-26 00:00:00 → 2025-07-05 00:00:00
refund_date: 2023-06-29 00:00:00 → 2025-07-09 00:00:00

Suppliers
created_date: 2022-06-24 00:00:00 → 2023-06-24 00:00:00


### Date Standardisation Findings

All identified date fields were successfully converted to datetime format with no invalid values.

Key observations:
- Customer registration dates range from June 2020 to June 2025.
- Product launch dates range from June 2021 to June 2025.
- Payment dates align with the main transaction period from June 2023 to June 2025.
- Return and refund activity extends into July 2025, which is reasonable because orders placed near the end of the analysis period may be returned or refunded after June 2025.
- No records were removed during date standardisation.

In [45]:
print(returns.columns.tolist())
print(returns.dtypes)

display(returns.head())

['return_id', 'order_id', 'customer_id', 'return_date', 'reason', 'return_amount', 'refund_status', 'refund_date', 'return_condition', 'remarks']
return_id                   object
order_id                    object
customer_id                 object
return_date         datetime64[ns]
reason                      object
return_amount              float64
refund_status               object
refund_date         datetime64[ns]
return_condition            object
remarks                     object
dtype: object


,return_id,order_id,customer_id,return_date,reason,return_amount,refund_status,refund_date,return_condition,remarks
0,RET00001,ORD000015,CUST05470,2024-12-23,Size Issue,24167.33,Pending,2024-12-30,Damaged,Customer initiated return within 7 day window
1,RET00002,ORD000017,CUST07581,2023-11-01,Defective Product,11698.27,Refunded,2023-11-06,Damaged,Customer initiated return within 7 day window
2,RET00003,ORD000038,CUST05691,2025-05-01,Wrong Product Delivered,11147.74,Refunded,2025-05-06,Damaged,Customer initiated return within 7 day window
3,RET00004,ORD000054,CUST03170,2025-06-20,Duplicate Order,9284.55,Refunded,2025-06-24,Acceptable,Customer initiated return within 7 day window
4,RET00005,ORD000072,CUST09318,2025-03-16,Duplicate Order,36912.59,Refunded,2025-03-23,Good,Customer initiated return within 7 day window


In [46]:
returns_check = returns.merge(
    orders[
        ['order_id', 'order_date', 'delivered_date', 'status']
    ],
    on='order_id',
    how='left',
    suffixes=('_return', '_order')
)

invalid_return_before_order = returns_check[
    returns_check['return_date'] < returns_check['order_date']
]

print(
    "Returns occurring before order date:",
    len(invalid_return_before_order)
)

Returns occurring before order date: 0


In [47]:
invalid_refund_dates = returns_check[
    returns_check['refund_date'] < returns_check['return_date']
]

print(
    "Refund dates before return dates:",
    len(invalid_refund_dates)
)

Refund dates before return dates: 0


In [48]:
return_before_delivery = returns_check[
    returns_check['delivered_date'].notna() &
    (returns_check['return_date'] < returns_check['delivered_date'])
]

print(
    "Returns recorded before delivery:",
    len(return_before_delivery)
)

Returns recorded before delivery: 1006


In [49]:
returns_without_delivery = returns_check[
    returns_check['delivered_date'].isna()
]

print(
    "Return records linked to orders without a delivery date:",
    len(returns_without_delivery)
)

print("\nTheir order statuses:")
print(
    returns_without_delivery['status']
    .value_counts(dropna=False)
)

Return records linked to orders without a delivery date: 5673

Their order statuses:
status
Returned     3933
Cancelled    1067
Shipped       673
Name: count, dtype: int64


### Returns Cross-Table Validation Findings

Return and refund records were validated against the corresponding order records.

- No returns occurred before the original order date.
- No refund dates occurred before their corresponding return dates.
- 1,006 return records have a return date earlier than the recorded delivery date.
- 5,673 return records are associated with orders that do not have a recorded delivery date.
- Among these 5,673 records:
  - 3,933 are marked as Returned
  - 1,067 are marked as Cancelled
  - 673 are marked as Shipped

These records indicate cross-table business-rule inconsistencies between the orders and returns datasets. They were flagged for data-quality reporting rather than automatically removed because the underlying business process or source-system logic should be confirmed before altering transactional records.

In [50]:
print("Total Return Records:", len(returns))
print("Unique Orders in Returns:", returns['order_id'].nunique())

duplicate_return_orders = (
    returns['order_id'].duplicated().sum()
)

print(
    "Additional return records for repeated order IDs:",
    duplicate_return_orders
)

Total Return Records: 10000
Unique Orders in Returns: 10000
Additional return records for repeated order IDs: 0


In [51]:
returns_per_order = (
    returns.groupby('order_id')
    .size()
    .value_counts()
    .sort_index()
)

print("Number of return records per order:")
print(returns_per_order)

Number of return records per order:
1    10000
Name: count, dtype: int64


### Return Record Uniqueness

- The returns table contains 10,000 return records representing 10,000 unique order IDs.
- No order appears more than once in the returns table.
- Therefore, duplicate return records are not responsible for the inconsistencies identified during return-date validation.

In [52]:
return_order_status = (
    returns[['order_id']]
    .merge(
        orders[['order_id', 'status']],
        on='order_id',
        how='left'
    )
)

print("Order Status for Orders Present in Returns Table:")
print(
    return_order_status['status']
    .value_counts(dropna=False)
)

Order Status for Orders Present in Returns Table:
status
Delivered    4327
Returned     3933
Cancelled    1067
Shipped       673
Name: count, dtype: int64


In [53]:
returned_status_orders = set(
    orders.loc[
        orders['status'] == 'Returned',
        'order_id'
    ]
)

return_table_orders = set(returns['order_id'])

print(
    "Orders with status Returned:",
    len(returned_status_orders)
)

print(
    "Orders present in returns table:",
    len(return_table_orders)
)

print(
    "Returned-status orders found in returns table:",
    len(returned_status_orders & return_table_orders)
)

print(
    "Returned-status orders missing from returns table:",
    len(returned_status_orders - return_table_orders)
)

Orders with status Returned: 3933
Orders present in returns table: 10000
Returned-status orders found in returns table: 3933
Returned-status orders missing from returns table: 0


In [54]:
print(
    "Return-table orders not marked Returned in orders:",
    len(return_table_orders - returned_status_orders)
)

Return-table orders not marked Returned in orders: 6067


### Order Status vs Returns Table Consistency

A cross-table comparison was performed between the orders and returns datasets.

- Total records in the returns table: 10,000
- Orders marked `Returned` in the orders table: 3,933
- All 3,933 orders marked as Returned have a corresponding record in the returns table.
- However, 6,067 additional orders in the returns table are not marked as Returned in the orders table.

Status distribution of orders appearing in the returns table:
- Delivered: 4,327
- Returned: 3,933
- Cancelled: 1,067
- Shipped: 673

This indicates a business-rule inconsistency between the order lifecycle status and the returns transaction table. The records were retained and flagged rather than modified because the correct source-of-truth definition should be confirmed before changing transactional data.

This inconsistency should also be considered when interpreting return-related KPIs.

In [55]:
order_status_summary = orders['status'].value_counts()

print("Order Status Distribution:")
print(order_status_summary)

print("\nPercentage Distribution:")
print(
    orders['status']
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

Order Status Distribution:
status
Delivered     32499
Cancelled      5996
Shipped        5071
Returned       3933
Processing     2501
Name: count, dtype: int64

Percentage Distribution:
status
Delivered     65.00
Cancelled     11.99
Shipped       10.14
Returned       7.87
Processing     5.00
Name: proportion, dtype: float64


In [56]:
expected_statuses = {
    'Delivered',
    'Cancelled',
    'Returned',
    'Shipped',
    'Processing'
}

actual_statuses = set(orders['status'].dropna().unique())

unexpected_statuses = actual_statuses - expected_statuses
missing_expected_statuses = expected_statuses - actual_statuses

print("Actual Statuses:", actual_statuses)
print("Unexpected Statuses:", unexpected_statuses)
print("Expected Statuses Not Present:", missing_expected_statuses)

Actual Statuses: {'Cancelled', 'Shipped', 'Returned', 'Delivered', 'Processing'}
Unexpected Statuses: set()
Expected Statuses Not Present: set()


In [57]:
cancelled_orders = (orders['status'] == 'Cancelled').sum()
total_orders = len(orders)

cancellation_rate = (
    cancelled_orders / total_orders
) * 100

print(f"Total Orders: {total_orders:,}")
print(f"Cancelled Orders: {cancelled_orders:,}")
print(f"Cancellation Rate: {cancellation_rate:.2f}%")

Total Orders: 50,000
Cancelled Orders: 5,996
Cancellation Rate: 11.99%


### Order Status and Cancellation Findings

The order status field contains only the five expected categories, with no unexpected or invalid status values.

Order status distribution:
- Delivered: 32,499 (65.00%)
- Cancelled: 5,996 (11.99%)
- Shipped: 5,071 (10.14%)
- Returned: 3,933 (7.87%)
- Processing: 2,501 (5.00%)

The overall cancellation rate is 11.99%, which is above the management benchmark of less than 10%.

This suggests a potential business risk from lost orders. Further analysis should investigate cancellation patterns by product category, payment method, geography, channel, and other relevant dimensions.

## Step 25: Numeric Business-Rule Validation

Key numeric fields were checked for impossible or suspicious values such as negative quantities, negative prices, negative transaction amounts, and invalid inventory levels.

These checks help ensure that business KPIs are not distorted by logically invalid numeric records.

In [58]:
print("ORDER ITEMS")

print("Quantity <= 0:",
      (order_items['quantity'] <= 0).sum())

print("Unit Price <= 0:",
      (order_items['unit_price'] <= 0).sum())

print("GST Amount < 0:",
      (order_items['gst_amount'] < 0).sum())

print("Discount Amount < 0:",
      (order_items['discount_amount'] < 0).sum())

print("Total Price <= 0:",
      (order_items['total_price'] <= 0).sum())

ORDER ITEMS
Quantity <= 0: 0
Unit Price <= 0: 0
GST Amount < 0: 0
Discount Amount < 0: 0
Total Price <= 0: 0


In [59]:
print("ORDERS")

print("Total Amount <= 0:",
      (orders['total_amount'] <= 0).sum())

print("GST Amount < 0:",
      (orders['gst_amount'] < 0).sum())

print("Shipping Charge < 0:",
      (orders['shipping_charge'] < 0).sum())

print("Discount Amount < 0:",
      (orders['discount_amount'] < 0).sum())

print("Final Amount <= 0:",
      (orders['final_amount'] <= 0).sum())

ORDERS
Total Amount <= 0: 0
GST Amount < 0: 0
Shipping Charge < 0: 0
Discount Amount < 0: 0
Final Amount <= 0: 0


In [60]:
print("PAYMENTS")

print("Payment Amount <= 0:",
      (payments['amount'] <= 0).sum())

print("Refund Amount < 0:",
      (payments['refund_amount'] < 0).sum())

PAYMENTS
Payment Amount <= 0: 0
Refund Amount < 0: 0


In [61]:
print("INVENTORY")

print("Quantity Available < 0:",
      (inventory['quantity_available'] < 0).sum())

print("Quantity Reserved < 0:",
      (inventory['quantity_reserved'] < 0).sum())

print("Reorder Level < 0:",
      (inventory['reorder_level'] < 0).sum())

print("Unit Cost <= 0:",
      (inventory['unit_cost'] <= 0).sum())

print("Inventory Value < 0:",
      (inventory['total_inventory_value'] < 0).sum())

INVENTORY
Quantity Available < 0: 0
Quantity Reserved < 0: 0
Reorder Level < 0: 0
Unit Cost <= 0: 0
Inventory Value < 0: 0


In [62]:
reserved_gt_available = inventory[
    inventory['quantity_reserved'] >
    inventory['quantity_available']
]

print(
    "Products where reserved quantity exceeds available quantity:",
    len(reserved_gt_available)
)

display(
    reserved_gt_available[
        [
            'inventory_id',
            'product_id',
            'quantity_available',
            'quantity_reserved',
            'status'
        ]
    ].head(10)
)

Products where reserved quantity exceeds available quantity: 0


,inventory_id,product_id,quantity_available,quantity_reserved,status


### Numeric Business-Rule Validation Findings

All major numeric fields passed the business-rule validation checks.

- No zero or negative product quantities were identified.
- No zero or negative unit prices or final order amounts were identified.
- No negative GST, discount, shipping, payment, or refund amounts were found.
- Inventory quantities, reorder levels, unit costs, and inventory values contain no invalid negative values.
- No products were found where reserved quantity exceeded available quantity.

Therefore, no numeric records required correction or removal based on these validation rules.

In [63]:
print("is_cod values:")
print(orders['is_cod'].value_counts(dropna=False))

print("\nPayment methods:")
print(orders['payment_method'].value_counts(dropna=False))

is_cod values:
is_cod
0    44117
1     5883
Name: count, dtype: int64

Payment methods:
payment_method
UPI                 17570
Credit Card         10007
Debit Card           7624
Cash on Delivery     5883
Net Banking          4026
EMI                  2943
Wallet               1947
Name: count, dtype: int64


In [64]:
cod_mismatch_1 = orders[
    (orders['payment_method'] == 'Cash on Delivery') &
    (orders['is_cod'] != 1)
]

cod_mismatch_2 = orders[
    (orders['payment_method'] != 'Cash on Delivery') &
    (orders['is_cod'] == 1)
]

print(
    "COD orders incorrectly flagged as non-COD:",
    len(cod_mismatch_1)
)

print(
    "Non-COD orders incorrectly flagged as COD:",
    len(cod_mismatch_2)
)

COD orders incorrectly flagged as non-COD: 0
Non-COD orders incorrectly flagged as COD: 0


In [65]:
inventory.groupby('status')[
    ['quantity_available', 'reorder_level']
].agg(['min', 'max', 'mean']).round(2)

quantity_available              reorder_level           
                            min  max    mean           min max   mean
status                                                               
In Stock                     15  500  256.46            10  50  29.54
Low Stock                     1   46   17.33            12  50  35.79
Out of Stock                  0    0    0.00            26  42  34.00

In [66]:
display(
    inventory[
        ['quantity_available', 'reorder_level', 'status']
    ].sort_values('quantity_available').head(20)
)

,quantity_available,reorder_level,status
853,0,42,Out of Stock
899,0,26,Out of Stock
565,1,45,Low Stock
971,1,37,Low Stock
299,1,27,Low Stock
114,2,31,Low Stock
248,2,46,Low Stock
654,2,31,Low Stock
672,4,48,Low Stock
827,4,30,Low Stock


### COD and Inventory Status Validation Findings

**Cash on Delivery Validation**
- 5,883 orders use Cash on Delivery.
- 5,883 orders are flagged with `is_cod = 1`.
- No COD orders were incorrectly flagged as non-COD.
- No non-COD orders were incorrectly flagged as COD.
- The COD indicator is fully consistent with the payment method field.

**Inventory Status Validation**
- 922 products are classified as In Stock.
- 76 products are classified as Low Stock.
- 2 products are classified as Out of Stock.
- Both Out of Stock products have zero available quantity.
- Low Stock products show substantially lower available quantities than products classified as In Stock.
- No obvious contradictions were identified between stock quantities and inventory status.

In [67]:
for name, df in datasets.items():
    print("\n" + "=" * 70)
    print(name.upper())
    print("=" * 70)
    print(df.dtypes)


CUSTOMERS
customer_id                  object
first_name                   object
last_name                    object
email                        object
phone                         int64
city                         object
state                        object
pincode                       int64
gender                       object
age                           int64
segment                      object
registration_date    datetime64[ns]
last_login_date      datetime64[ns]
is_verified                   int64
is_active                     int64
total_orders                  int64
total_spent                 float64
dtype: object

ORDERS
order_id                    object
customer_id                 object
order_date          datetime64[ns]
order_time                  object
status                      object
city                        object
state                       object
pincode                      int64
total_amount               float64
gst_amount                 float64
shipp

### Final Data-Type Audit Findings

The final data-type audit confirms that all required date fields were successfully converted to datetime format.

The only exception is `payments.refund_date`, which remains non-datetime because the column contains 100% missing values and no payment record contains a refund amount.

No additional data-type corrections are required at this stage.

In [69]:
quality_summary = pd.DataFrame({
    'Check': [
        'Complete duplicate rows',
        'Duplicate order IDs',
        'Invalid order dates',
        'Delivery before order date',
        'Delivered orders missing delivery date',
        'Orders above ₹5 lakh',
        'IQR final_amount outliers',
        'Payment failure rate',
        'Order-item orphan records',
        'Payment orphan records',
        'Return orphan records',
        'Return before order date',
        'Return before delivery',
        'Returns linked to orders without delivery date',
        'Return-table orders not marked Returned',
        'Out-of-stock products',
        'Customers with zero orders',
        'Customer total_orders mismatches',
        'COD flag mismatches',
        'Invalid numeric values'
    ],
    
    'Result': [
        '0',
        '0',
        '0',
        '0',
        '0',
        '312',
        '5,243 (10.49%)',
        '3.50%',
        '0',
        '0',
        '0',
        '0',
        '1,006',
        '5,673',
        '6,067',
        '2',
        '457 (4.57%)',
        '0',
        '0',
        '0'
    ],
    
    'Assessment': [
        'PASS',
        'PASS',
        'PASS',
        'PASS',
        'PASS',
        'REVIEW',
        'REVIEW',
        'BUSINESS RISK',
        'PASS',
        'PASS',
        'PASS',
        'PASS',
        'DATA QUALITY ISSUE',
        'DATA QUALITY ISSUE',
        'DATA QUALITY ISSUE',
        'INFO',
        'INFO',
        'PASS',
        'PASS',
        'PASS'
    ]
})

display(quality_summary)

,Check,Result,Assessment
0,Complete duplicate rows,0,PASS
1,Duplicate order IDs,0,PASS
2,Invalid order dates,0,PASS
3,Delivery before order date,0,PASS
4,Delivered orders missing delivery date,0,PASS
5,Orders above ₹5 lakh,312,REVIEW
6,IQR final_amount outliers,"5,243 (10.49%)",REVIEW
7,Payment failure rate,3.50%,BUSINESS RISK
8,Order-item orphan records,0,PASS
9,Payment orphan records,0,PASS


In [68]:
date_audit = {
    'customers.registration_date': customers['registration_date'].dtype,
    'customers.last_login_date': customers['last_login_date'].dtype,
    'orders.order_date': orders['order_date'].dtype,
    'orders.delivered_date': orders['delivered_date'].dtype,
    'products.launch_date': products['launch_date'].dtype,
    'payments.payment_date': payments['payment_date'].dtype,
    'returns.return_date': returns['return_date'].dtype,
    'returns.refund_date': returns['refund_date'].dtype,
    'inventory.last_restocked_date': inventory['last_restocked_date'].dtype,
    'suppliers.created_date': suppliers['created_date'].dtype
}

for column, dtype in date_audit.items():
    print(f"{column}: {dtype}")

customers.registration_date: datetime64[ns]
customers.last_login_date: datetime64[ns]
orders.order_date: datetime64[ns]
orders.delivered_date: datetime64[ns]
products.launch_date: datetime64[ns]
payments.payment_date: datetime64[ns]
returns.return_date: datetime64[ns]
returns.refund_date: datetime64[ns]
inventory.last_restocked_date: datetime64[ns]
suppliers.created_date: datetime64[ns]


### Final Date Audit Result

All required date columns were successfully converted to `datetime64[ns]`.

Validated date fields include:
- Customer registration and last login dates
- Order and delivery dates
- Product launch dates
- Payment dates
- Return and refund dates
- Inventory restock dates
- Supplier creation dates

No further date-type corrections are required.

# Phase 1 — Data Quality Report

## Executive Summary

The IndiaKart dataset contains 8 interconnected tables covering customers, orders, order items, products, payments, returns, inventory, and suppliers. A total of 222,200 records were reviewed during the data quality assessment.

Overall, the dataset is structurally strong. No duplicate rows, duplicate primary identifiers, orphan records, invalid order dates, or invalid numeric values were identified. All relevant date fields were standardized to datetime format, and key relationships between customers, orders, products, payments, and returns were successfully validated.

However, several cross-table inconsistencies were identified in the returns data and should be considered when interpreting return-related business metrics.

## Key Data Quality Findings

### 1. Missing Values
Six of the eight datasets contain no missing values.

The `orders.delivered_date` field contains 17,501 missing values (35%). These values were retained because they occur among non-delivered order statuses and can represent valid business conditions.

The `payments.refund_date` field is entirely blank. Investigation showed that `refund_amount` is ₹0 across all 50,000 payment records, indicating that no refund transactions are recorded in the payments table.

### 2. Duplicate and Referential Integrity Checks
No completely duplicated records were identified across any of the eight datasets.

All tested primary identifiers are unique, including customer, order, item, product, payment, return, and supplier IDs.

All tested relationships passed referential integrity validation. No orphan records were found between order items and orders, payments and orders, returns and orders, orders and customers, or order items and products.

### 3. Order Value Review
A total of 312 orders have a final value above ₹5,00,000 and were flagged for review.

Using the IQR method, 5,243 orders (10.49%) were identified as statistical outliers. These transactions were retained because high-value orders may represent legitimate purchases rather than data errors.

### 4. Returns Data Quality Issues
Returns data contains the most significant consistency issues identified during Phase 1.

- 1,006 return records have a return date earlier than the recorded delivery date.
- 5,673 return records are linked to orders without a recorded delivery date.
- 6,067 orders appearing in the returns table are not marked as `Returned` in the orders table.
- All 3,933 orders marked as `Returned` in the orders table do have corresponding records in the returns table.

These records were flagged rather than modified because the correct source-of-truth business rules should be confirmed before altering transactional data.

## Business Risk Indicators

Two early business risks were identified during the quality assessment:

- Payment Failure Rate: 3.50%, above the management target of less than 2%.
- Cancellation Rate: 11.99%, above the management target of less than 10%.

These are valid business events rather than data-cleaning errors and should be investigated further during exploratory and KPI analysis.

Inventory currently contains 2 Out of Stock products and 76 Low Stock products. In addition, 457 customers (4.57%) have placed zero orders, which is considered valid because registered customers may not yet have completed a purchase.

## Cleaning Decisions

No rows were deleted during Phase 1.

Date fields were standardized, high-value transactions were flagged rather than removed, logically valid missing values were retained, and identified returns inconsistencies were documented for further investigation.

The cleaned datasets are suitable for exploratory analysis, with caution required when interpreting return-related metrics because of the identified cross-table inconsistencies.

In [71]:
clean_path = r"C:\Users\karti\Desktop\IndiaEKART\cleaned_data"

os.makedirs(clean_path, exist_ok=True)

customers.to_csv(
    os.path.join(clean_path, "customers_clean.csv"),
    index=False
)

orders.to_csv(
    os.path.join(clean_path, "orders_clean.csv"),
    index=False
)

order_items.to_csv(
    os.path.join(clean_path, "order_items_clean.csv"),
    index=False
)

products.to_csv(
    os.path.join(clean_path, "products_clean.csv"),
    index=False
)

payments.to_csv(
    os.path.join(clean_path, "payments_clean.csv"),
    index=False
)

returns.to_csv(
    os.path.join(clean_path, "returns_clean.csv"),
    index=False
)

inventory.to_csv(
    os.path.join(clean_path, "inventory_clean.csv"),
    index=False
)

suppliers.to_csv(
    os.path.join(clean_path, "suppliers_clean.csv"),
    index=False
)

print("All cleaned datasets saved successfully.")

All cleaned datasets saved successfully.


In [72]:
clean_files = os.listdir(clean_path)

print("Cleaned files created:")
for file in clean_files:
    print(file)

print("\nTotal cleaned files:", len(clean_files))

Cleaned files created:
customers_clean.csv
inventory_clean.csv
orders_clean.csv
order_items_clean.csv
payments_clean.csv
products_clean.csv
returns_clean.csv
suppliers_clean.csv

Total cleaned files: 8
